### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_snmp_user.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_snmp_user.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [config snmp user] ********************************************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_snmp_user.yml:27
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_snmp_user.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378 `" && echo ansible-tmp-1487053788.71-134455651982378="` echo $HOME/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpgWgzQ_ TO /root/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378/ce_snmp_user.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378/ /root/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378/ce_snmp_user.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378/ce_snmp_user.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487053788.71-134455651982378/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "snmp local user": {
            "local_user_info": []
        }, 
        "snmp usm user": {
            "usm_user_info": [
                {
                    "aclNumber": "2000", 
                    "engineID": "800007DB03FCE33CA7D461", 
                    "remoteEngineID": "false", 
                    "userName": "wdz_snmp"
                }
            ]
        }
    }, 
    "existing": {
        "snmp local user": {
            "local_user_info": []
        }, 
        "snmp usm user": {
            "usm_user_info": []
        }
    }, 
    "invocation": {
        "module_args": {
            "aaa_local_user": null, 
            "acl_number": "2000", 
            "auth_key": null, 
            "auth_pass": null, 
            "auth_protocol": null, 
            "authorize": false, 
            "host": "ce_6850", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "priv_key": null, 
            "priv_protocol": null, 
            "provider": null, 
            "remote_engine_id": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "user_group": null, 
            "username": "rootDC", 
            "usm_user_name": "wdz_snmp", 
            "validate_certs": true
        }, 
        "module_name": "ce_snmp_user"
    }, 
    "proposed": {
        "acl_number": "2000", 
        "state": "present", 
        "usm_user_name": "wdz_snmp"
    }, 
    "updates": [
        [
            "snmp-agent usm-user v3 wdz_snmp acl 2000", 
            "snmp-agent usm-user v3 wdz_snmp", 
            "snmp-agent usm-user v3 wdz_snmp"
        ]
    ]
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 1]
```diff
@@ -0,0 +1,1074 @@
...
+#
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: ce_snmp_user
+version_added: "2.3"
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: `version_added: "2.4"`

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 2]
```diff
@@ -0,0 +1,1074 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: `metadata_version`

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 3]
```diff
@@ -0,0 +1,1074 @@
...
+        required: false
+        default: null
+    user_group:
+        description:
+            - Name of the group where user belongs to.
+        required: false
+        default: null
+    auth_protocol:
+        description:
+            - Authentication protocol ( md5 | sha ).
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: `Authentication protocol.`

No need to repeat `choices:`

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 4]
```diff
@@ -0,0 +1,1074 @@
...
+        required: false
+        default: null
+        choices: ['noAuth', 'md5', 'sha']
+    auth_key:
+        description:
+            - The authentication password. Simple password length <8-255>. Field max.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: Suggestion: `The authentication password. Password length, 8-255 characters.`

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 5]
```diff
@@ -0,0 +1,1074 @@
...
+        default: null
+    priv_protocol:
+        description:
+            - Encryption protocol.
+        required: false
+        default: null
+        choices: ['noPriv', 'des56', '3des168', 'aes128', 'aes192', 'aes256']
+    priv_key:
+        description:
+            - The encryption password. Simple password length <8-255>. Field max.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: Suggestion: `The encryption password. Password length 8-255 characters.`

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 6]
```diff
@@ -0,0 +1,1074 @@
...
+        auth_protocol = module.params['auth_protocol']
+        auth_key = module.params['auth_key']
+        priv_protocol = module.params['priv_protocol']
+        priv_key = module.params['priv_key']
+
+        local_user_name = module.params['aaa_local_user']
+
+        if usm_user_name:
+
+            if local_user_name:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
```

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py [hunk 7]
```diff
@@ -0,0 +1,1074 @@
...
+
+            if auth_key:
+                if len(auth_key) > 255 or len(auth_key) == 0:
+                    module.fail_json(
+                        msg='Error: The length of auth_key %s is out of [1 - 255].' % auth_key)
+
+                conf_str += "<authKey></authKey>"
+
+            if priv_protocol:
+                if not auth_protocol:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_snmp_user.py:390]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_together
require_if
```

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 2 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: OK, I'll fix it. thanks.

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: OK, I'll fix it. thanks.

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 5 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: Good suggestion, I'll fix it. thanks.

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 7 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py:390]: I'm trying, but it not work well for this logic

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 6 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py]: OK, I'll fix it, thanks.

File: lib/ansible/modules/network/cloudengine/ce_snmp_user.py (hunk 7 above)
Comment (gundalow) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py:390]: I'm currently working on some documentation for `AnsibleModule`, the following shows the different validation options:
```
def main():
    module = AnsibleModule(
        argument_spec=dict(
            state=dict(default='present', choices=['present', 'absent']),
            username=dict(type='str'),
            password=dict(no_log=True),
            token=dict(no_log=True),
            src=dict(type='path'),
            priority=dict(type='int'),
        ),
        mutually_exclusive=(['password', 'token'],),
        required_together=(['username', 'password'],),
        required_one_of=[['password', 'token']],
        required_if=[('state', 'present', ['src', 'priority'])],
        supports_check_mode=True
   )
```

Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_snmp_user.py:390]: @gundalow I need your help. when I deal this logic:
```
if A:
    if not B:
        print something errror
```
B is not dependent on A, but A dependent on B， They don't need to be config at the same time， it's not use ```required_together```, so I'll try:
```
required_if=[('B', not None, ['A'])]
```
but it is not work well.
I think ```required_if``` needs to be improved, This condition is not enough to judge when the condition is ```not None```,  How do you think about?

**Comment by QijunPan** [comment]
@gundalow   These problems have been fixed， can  this PR merged?

**Review by gundalow — APPROVED** [review]

**Pull request merged**